### PR TITLE
added test for pair_programming in Git backend

### LINKED ISSUE #825: pair_programming for git backend doesn't work right
So I was exploring the Git enricher and came through this issue with  `pair_programming` set to `true` in `setup_cfg`.

The variables `m` and `n` set [here](https://github.com/chaoss/grimoirelab-elk/blob/8ea3455bd639de695e1a4d2f56ed86df4e5818c7/grimoire_elk/enriched/git.py#L414) through regex don't seem to work for the repository with multi-authored commits I've chosen (https://github.com/juliettepretot/npx-co-commit).

Based on the current logic both the variables are set to `None` because the regex used in `AUTHOR_P2P_REGEX` and `AUTHOR_P2P_NEW_REGEX` seems faulty, when they both should get populated by the 2 author names for that particular commit.

For instance let's look at this commit JSON data from Perceval 
```
{
    "backend_name": "Git",
    "backend_version": "0.12.0",
    "category": "commit",
    "classified_fields_filtered": null,
    "data": {
        "Author": "Juliette Pretot <hi@juliette.sh>",
        "AuthorDate": "Mon Oct 29 16:29:48 2018 +0100",
        "Commit": "Juliette Pretot <hi@juliette.sh>",
        "CommitDate": "Tue Nov 6 21:41:35 2018 +0100",
        "commit": "dbc4e2c18c893be4db75b472d4016e623316729e",
        "files": [
            {
                "action": "A",
                "added": "10",
                "file": ".gitignore",
                "indexes": [
                    "0000000",
                    "e45672e"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "24",
                "file": "README.MD",
                "indexes": [
                    "0000000",
                    "45d1af5"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "46",
                "file": "index.js",
                "indexes": [
                    "0000000",
                    "5f2990c"
                ],
                "modes": [
                    "000000",
                    "100755"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "27",
                "file": "package-lock.json",
                "indexes": [
                    "0000000",
                    "d0ef368"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "19",
                "file": "package.json",
                "indexes": [
                    "0000000",
                    "8f521a3"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "-",
                "file": "screenshot.gif",
                "indexes": [
                    "0000000",
                    "74de826"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "-"
            }
        ],
        "message": "initial commit\n\nCo-authored-by: mariiapunda <mariiapunda@users.noreply.github.com>",
        "parents": [],
        "refs": []
    },
    "origin": "https://github.com/juliettepretot/npx-co-commit",
    "perceval_version": "0.12.24",
    "search_fields": {
        "item_id": "dbc4e2c18c893be4db75b472d4016e623316729e"
    },
    "tag": "https://github.com/juliettepretot/npx-co-commit",
    "timestamp": 1585582796.576738,
    "updated_on": 1541536895.0,
    "uuid": "0c9876869058dc5b3f836365f3ce06a5b5575749"
}
```
According to the above commit data there must be 2 authors for the commit **Juliette Pretot** and **mariiapunda**  (based on the line `"message": "initial commit\n\nCo-authored-by: mariiapunda `)

But after fixing regex 
```
AUTHOR_P2P_REGEX = re.compile(r'(?P<first_authors>.* .*) and (?P<last_author>.* .*) (?P<email>.*)')
```
 to
```
AUTHOR_P2P_REGEX = re.compile(r'(?P<first_authors>.* ) (?P<last_author> .*) (?P<email>.*)')
```
what I observe is in the function `__get_authors()` [here](https://github.com/chaoss/grimoirelab-elk/blob/8ea3455bd639de695e1a4d2f56ed86df4e5818c7/grimoire_elk/enriched/git.py#L118) the committer name from this line 

```
"Author": "Juliette Pretot <hi@juliette.sh>",
```

is used and gets split into two names and saved in `item['data']['authors']` as `<class 'list'>: ['Juliette', 'Pretot']`.

AFAIK `item['data']['authors']` in this case should be `<class 'list'>: ['Juliette Pretot', 'mariiapunda']`.

I feel the logic being used here is wrong, from the regex to how multiple authors are being identified.
Please tell me if I am wrong with the repository I chose or the way I am interpreting the logic.

### PR DESCRIPTION
**Pull request by imnitishng**
Added tests for `pair_programming` in Git backend to increase coverage
Fixes https://github.com/chaoss/grimoirelab-elk/issues/825

Signed-off-by: Nitish Gupta <imnitish.ng@gmail.com>

### CODE REVIEW

File: tests/base.py [hunk 1]
```diff
@@ -274,10 +274,18 @@ def _test_study(self, test_study):
         data2es(self.items, self.ocean_backend)
 
         # populate enriched index
-        self.enrich_backend = self.connectors[self.connector][2](db_sortinghat=DB_SORTINGHAT,
-                                                                 db_user=self.db_user,
-                                                                 db_password=self.db_password,
-                                                                 db_projects_map=self.db_projects_map)
+        # different arguments to test pair_programming
```
Comment (valeriocos) [inline tests/base.py:277]: Why you want to test the pair programming with the demography study? 

Pair programming could be calculated in a different test, leveraging on the _test_raw_to_enrich method.

File: tests/data/git.json [hunk 2]
```diff
@@ -61,16 +61,20 @@
     "tag": "/tmp/perceval_mc84igfc/gittest",
     "timestamp": 1518165202.694393,
     "updated_on": 1344965413.0,
+    "git_uuid": "43f217b2f678a5691fdbc5c6c5302243e79e5a90",
```
Comment (valeriocos) [inline tests/data/git.json:64]: why is it needed to include the attribute `git_uuid`? This comment applies also to the other changes on this attribute

**Review by valeriocos — COMMENTED** [review]
Thank you @imnitishng for the PR. The approach looks OK, however I would implement a specific test case for the Git pair programming. The main reason is because with the current implementation we are not testing whether the pair programming attributes are correctly saved to the enriched indexes. Furthermore, we are coupling the testing of the pair programming to the demography study, which doesn't seem to be needed.

Please have a look at the proposal below and share your thoughts. Thanks!

The tests/base.py could be modified as follow:
```
--- a/tests/base.py
+++ b/tests/base.py
@@ -162,7 +162,7 @@ class TestBaseBackend(unittest.TestCase):
 
         return {'items': len(self.items), 'raw': raw_items}
 
-    def _test_raw_to_enrich(self, sortinghat=False, projects=False):
+    def _test_raw_to_enrich(self, sortinghat=False, projects=False, pair_programming=False):
         """Test whether raw indexes are properly enriched"""
 
         # populate raw index
@@ -185,6 +185,9 @@ class TestBaseBackend(unittest.TestCase):
                                                                      db_user=self.db_user,
                                                                      db_password=self.db_password)
 
+        if pair_programming:
+            self.enrich_backend.pair_programming = pair_programming
+
         elastic_enrich = get_elastic(self.es_con, self.enrich_index, clean, self.enrich_backend, self.enrich_aliases)
         self.enrich_backend.set_elastic(elastic_enrich)

```

We could keep only the modifications below of the Git test data:
```
--- a/tests/data/git.json
+++ b/tests/data/git.json
@@ -67,10 +67,13 @@
     "backend_version": "0.8.10",
     "category": "commit",
     "data": {
-        "Author": "Eduardo Morais <companheiro.vermelho@gmail.com>",
+        "Author": "Eduardo Morais and Zhongpeng Lin <companheiro.vermelho@gmail.com>",
         "AuthorDate": "Tue Aug 14 14:32:15 2012 -0300",
-        "Commit": "Eduardo Morais <companheiro.vermelho@gmail.com>",
+        "Commit": "Eduardo Morais and Zhongpeng Lin <companheiro.vermelho@gmail.com>",
         "CommitDate": "Tue Aug 14 14:32:15 2012 -0300",
+        "Signed-off-by": [
+            "Eduardo Morais <companheiro.vermelho@gmail.com>"
+        ],
         "commit": "87783129c3f00d2c81a3a8e585eb86a47e39891a",
         "files": [
             {
@@ -89,7 +92,7 @@
                 "removed": "0"
             }
         ],
-        "message": "Renamed file",
+        "message": "Renamed file.\\\\n\\\\nSigned-off-by: David Calavera <david.calavera@gmail.com>",
         "parents": [
             "bc57a9209f096a130dcc5ba7089a8663f758a703"
         ],
```

The tests could be modified as follows:
```
--- a/tests/test_git.py
+++ b/tests/test_git.py
@@ -88,9 +88,9 @@ class TestGit(TestBaseBackend):
         self.assertEqual(eitem['utc_author_date_weekday'], 2)
         self.assertEqual(eitem['utc_author_date_hour'], 17)
 
-        self.assertEqual(eitem['author_uuid'], 'f3aee5067d4691544f10d915932c9f1d08cb3b36')
+        self.assertEqual(eitem['author_uuid'], '8b8d552af706acff79df0f18f5295391c51acd79')
         self.assertEqual(eitem['author_domain'], 'gmail.com')
-        self.assertEqual(eitem['author_name'], 'Eduardo Morais')
+        self.assertEqual(eitem['author_name'], 'Eduardo Morais and Zhongpeng Lin')
 
         self.assertEqual(eitem['commit_date'], '2012-08-14T14:32:15')
         self.assertEqual(eitem['commit_date_weekday'], 2)
@@ -119,6 +119,20 @@ class TestGit(TestBaseBackend):
         aliases = self.enrich_backend.elastic.list_aliases()
         self.assertListEqual(self.enrich_aliases, list(aliases.keys()))
 
+    def test_raw_to_enrich_pair_programming(self):
+        """Test whether the raw index is properly enriched with pair programming info"""
+
+        result = self._test_raw_to_enrich(pair_programming=True)
+        self.assertEqual(result['raw'], 9)
+        self.assertEqual(result['enrich'], 11)
+
+        enrich_backend = self.connectors[self.connector][2]()
+        url = self.es_con + "/" + self.enrich_index + "/_search"
+        response = enrich_backend.requests.get(url, verify=False).json()
+
+        for hit in response['hits']['hits']:
+            print("here") <--- TBD
+
     def test_enrich_repo_labels(self):
         """Test whether the field REPO_LABELS is present in the enriched items"""
 
@@ -403,8 +417,8 @@ class TestGit(TestBaseBackend):
         self.assertEqual(item['Commit'], '')
 
         item = self.items[1]['data']
-        self.assertEqual(item['Author'], 'e2ea52f7f782fe08109b762e474ff20656a51f47 <xxxxxx@gmail.com>')
-        self.assertEqual(item['Commit'], 'e2ea52f7f782fe08109b762e474ff20656a51f47 <xxxxxx@gmail.com>')
+        self.assertEqual(item['Author'], '7bb272958a7c0c54de85dc078aa1b98da7b930de <xxxxxx@gmail.com>')
+        self.assertEqual(item['Commit'], '7bb272958a7c0c54de85dc078aa1b98da7b930de <xxxxxx@gmail.com>')
 
         item = self.items[6]['data']
         self.assertEqual(item['Author'], 'abe1a5515d468ed258124c4c946ceb34ef7ffbda <xxxxxx@gmail.com>')
@@ -431,11 +445,11 @@ class TestGit(TestBaseBackend):
 
         item = self.items[1]
         eitem = enrich_backend.get_rich_item(item)
-        self.assertEqual(eitem['author_uuid'], '5fe609e6ddadd19697e69832ac1889bb942cccfb')
+        self.assertEqual(eitem['author_uuid'], '0442af9edda128740e57331769ee9e68a99ec36f')
         self.assertEqual(eitem['author_domain'], 'gmail.com')
-        self.assertEqual(eitem['author_name'], 'e2ea52f7f782fe08109b762e474ff20656a51f47')
+        self.assertEqual(eitem['author_name'], '7bb272958a7c0c54de85dc078aa1b98da7b930de')
         self.assertEqual(eitem['committer_domain'], 'gmail.com')
-        self.assertEqual(eitem['committer_name'], 'e2ea52f7f782fe08109b762e474ff20656a51f47')
+        self.assertEqual(eitem['committer_name'], '7bb272958a7c0c54de85dc078aa1b98da7b930de')
 
         item = self.items[6]
         eitem = enrich_backend.get_rich_item(item)
```

File: tests/data/git.json (hunk 2 above)
Comment (imnitishng) [reply tests/data/git.json:64]: The `git_uuid` attribute is created [here](https://github.com/chaoss/grimoirelab-elk/blob/39596eae4441815fda65ac8374e7fb4cd8781088/grimoire_elk/enriched/git.py#L110)) once `pair_programming` parameter is set to True in the GitEnrich class. 
This function is called several times while uploading data to elasticsearch, [for instance here](https://github.com/chaoss/grimoirelab-elk/blob/39596eae4441815fda65ac8374e7fb4cd8781088/grimoire_elk/enriched/enrich.py#L1553) so the attribute was needed.

File: tests/base.py (hunk 1 above)
Comment (imnitishng) [reply tests/base.py:277]: Okay. I get the point. I am sorry I did not realize this, I will soon rework this and send the PR.

Comment (valeriocos) [reply tests/base.py:277]: No worries, thank you @imnitishng :)

File: tests/data/git.json (hunk 2 above)
Comment (valeriocos) [reply tests/data/git.json:64]: I see, thank you for the pointers. 

The test data is basically the data that comes from Perceval, and it is modified to reflect data that comes directy from Perceval (e.g., a missing attribute, a multi authored commit). The `git_uuid` is something that is added during the enrichment phase, thus it shouldn't appear in the test data.

Let me know if you have questions, thanks!

Comment (imnitishng) [reply tests/data/git.json:64]: Okay I get it. Thank You! But while working on this PR I saw absence of this attribute caused some tests to fail. Nevermind I will work on the new approach and get to you if I encounter any problems or doubts related to this attribute.

Comment (valeriocos) [reply tests/data/git.json:64]: Ok, thanks!

**Comment by valeriocos** [comment]
please @imnitishng ping me when the PR is ready for review (or let me know if you need some help), thanks!

**Comment by imnitishng** [comment]
Okay, @valeriocos, sorry for the delay I have been busy with gitter backend, I will send this PR tomorrow.

**Comment by valeriocos** [comment]
no worries @imnitishng , thank you for the quick reply!

**Comment by imnitishng** [comment]
Hi @valeriocos, please have a look I have fixed the `pair_programming` issue, however 2 new demographic tests for `gitter` and `git` have been failing, I tried but couldn't find the issue, because the files I have changed don't seem to affect them. I think the issue is with my machine, so I'll wait and check if the tests here pass successfully.

Screenshot from 2020-04-09 23-23-55

**Comment by imnitishng** [comment]
Okay, checks seem to pass here, IDK why they failed on my machine, anyway please review the PR when you get the time.

**Comment by imnitishng** [comment]
Okay, checks seem to pass here, IDK why they failed on my machine, anyway please review the PR when you get the time.

**Comment by valeriocos** [comment]
The tests fail because thre is already an index in your machine using the demographics alias. If you remove that alias (or delete the index) , the tests will pass.

**Review by valeriocos — COMMENTED** [review]
That's a really nice contribution, thank you @imnitishng . Can you squash the commits into one and summarize the changes in the title and commit description?

Thanks!

**Comment by imnitishng** [comment]
@valeriocos, please have look.

**Review by valeriocos — APPROVED** [review]
LGTM, thanks @imnitishng

**Pull request merged**